source/agora/consensus/ValidatorSet.d

```diff
@@ -436,86 +406,88 @@ public class ValidatorSet
     {
-        static ubyte[] buffer;
-
-        // check if the enrollment data exists
-        Enrollment stored_enroll;
         try
         {
-            if (!this.getEnrollment(preimage.enroll_key, stored_enroll))
-            {
-                log.info("Rejected adding a pre-image for non-existing " ~
-                    "enrollment, Preimage: {}", preimage);
-                return false;
-            }
-        }
-        catch (Exception ex)
-        {
-            log.error("Database operation error on addPreimage: {}, " ~
-                "Preimage: {}", ex.msg, preimage);
-            return false;
-        }
+            auto results = this.db.execute("SELECT preimage, distance FROM " ~
+                "validator_set WHERE key = ? AND distance < ?",
+                enroll_hash.toString(), new_distance);
 
-        // check if already exists
-        try
-        {
-            if (this.hasPreimage(preimage.enroll_key, preimage.distance))
-            {
-                log.info("Rejected already existing pre-image, Preimage: {}",
-                    preimage);
+            if (results.empty || results.oneValue!(byte[]).length == 0)
                 return false;
-            }
+
+            auto row = results.front;
+            preimage = Hash(row.peek!(char[])(0));
+            distance = row.peek!ushort(1);
+            return true;
         }
         catch (Exception ex)
         {
-            log.error("Database operation error on addPreimage: {}, " ~
-                "Preimage: {}", ex.msg, preimage);
+            log.error("Error calling getOlderPreimageDistance for key {}: {}",
+                enroll_hash, ex);
             return false;
         }
+    }
 
-        // check the validity of new pre-image based on the stored pre-image
-        PreImageInfo stored_image = this.getPreimage(preimage.enroll_key);
-        if (stored_image != PreImageInfo.init)
-        {
-            if (auto reason = isInvalidReason(
-                preimage.distance, preimage.hash, stored_image.distance,
-                stored_image.hash, Enrollment.ValidatorCycle))
-            {
-                log.info("Invalid preimage data: {}, Data was: ", reason,
-                    preimage);
-                return false;
-            }
-        }
+    /***************************************************************************
 
-        // insert the pre-image into the table
-        try
+        Add a pre-image information to a validator data
+
+        Params:
+            preimage = the pre-image information to add
+
+        Returns:
+            true if the pre-image information has been added to the validator
+
+    ***************************************************************************/
+
+    public bool addPreimage (const ref PreImageInfo preimage) @trusted nothrow
+    {
+        Hash prev_preimage;
+        ushort prev_distance;
+        if (!this.getOlderPreimageDistance(preimage.enroll_key,
```

Comment:
Maybe we don't need this. Maybe we could just add a `new_distance` parameter to `getPreimage`. It would simplify the implementation.